external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp

```diff
@@ -513,54 +554,159 @@ class TosaPartitionPass : public TosaPartitionBase<TosaPartitionPass> {
-        outlineConvPartOps(convOp, secondOps.getArrayRef(), frontOps,
-                           inputNodes.getArrayRef(), resultNodes.getArrayRef(),
+        outlineConvPartOps(anchorOp, trailingOps.getArrayRef(),
+                           leadingOps.getArrayRef(), inputNodes.getArrayRef(),
+                           resultNodes.getArrayRef(),
                            std::string(func.getSymName()) + strCount,
-                           candidates);
+                           attributeName(), candidates);
         // Outlining will erase nodes and thus perturb the walk, so
         // signal interrupted to exit it and restart.
         return WalkResult::interrupt();
       };
 
-      // Walk until we've outlined all the Conv2D ops we can.
+      // Walk until we've outlined all the anchor ops we can.
       while (func.walk(callback).wasInterrupted()) {
       }
     }
   }
 };
 
+class TosaPartitionPassWithOptions
+  : public TosaPartitionPass {
+public:
+  TosaPartitionPassWithOptions() = default;
+  TosaPartitionPassWithOptions(ArrayRef<std::string> anchorOps_,
+                               std::string attrName, bool trailingOnly_) {
+    anchorOps = anchorOps_;
+    attributeNameOpt = attrName;
+    trailingOnly = trailingOnly_;
+  }
+
+  bool isAnchorOp(Operation *op) override {
+    if (anchorOps.empty())   // ListOption doesn't have a default value.
+      anchorOps = {"tosa.conv2d","tosa.matmul","tosa.depthwise_conv2d"};
+
+    return llvm::is_contained(anchorOps, op->getName().getIdentifier().str());
+  }
+  bool isLeadingOp(Operation *op) override {
+    return !trailingOnly && (isConstantZero(op) || isFuseableOp(op));
+  }
+  std::string attributeName() override {
+    return attributeNameOpt;
+  }
+};
+
 } // namespace
 
 std::unique_ptr<Pass> mlir::tosa::createTosaPartitionPass() {
-  return std::make_unique<TosaPartitionPass>();
+  return std::make_unique<TosaPartitionPassWithOptions>();
+}
+
+namespace {
+
+class TestTosaPartitionOptionsPass
+    : public PassWrapper<TestTosaPartitionOptionsPass,
+                         OperationPass<ModuleOp>> {
+public:
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(TestTosaPartitionOptionsPass)
+
+  StringRef getArgument() const final { return "test-tosa-partition-options"; }
+  StringRef getDescription() const final {
+    return "Tests the programmatic interface to --tosa-partition options.";
+  }
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<tosa::TosaDialect>();
+  }
+
+  TestTosaPartitionOptionsPass() = default;
+  TestTosaPartitionOptionsPass(const TestTosaPartitionOptionsPass &) {}
+
+  void runOnOperation() override {
+    ModuleOp module = getOperation();
+    PassManager pm(module.getContext(), mlir::PassManager::Nesting::Implicit);
+    if (defaultCase) {
+      pm.addPass(tosa::createTosaPartitionPass());
+    } else if (depthwiseOnly) {
+      class DepthwiseOnlyPartitionPass
```

Comment:
The last version made specialised PartitionConfig subclasses, where this version makes specialised TosaPartitionPass subclasses.